src/userSurvey.js

```diff
@@ -1,118 +1,154 @@
-const surveyRequestTitle = "New Ion survey available";
+let createdSurvey = false;
 
-const secondsPerDay = 86400;
+// Module-wide variables for a survey, set in createSurvey
+let lastSurveyRequest = 0;
+let reminderIconUrl = "";
+let reminderInterval = 0;
+let reminderMessage = "";
+let reminderTitle = "";
+let surveyUrl = "";
 
 const millisecondsPerSecond = 1000;
 
-const surveyRemindPeriodDays = 3;
-
-const surveyCompletionUrl = "https://citpsurveys.cs.princeton.edu/thankyou";
-
-let surveyUrlBase = "";
-
 /**
- * Generates a RFC4122 compliant ID
- * https://www.ietf.org/rfc/rfc4122.txt based on given seed.
- *
- * A compliant UUID is of the form
- * xxxxxxxx-xxxx-Mxxx-Nxxx-xxxxxxxxxxxx; where 1 <= M <= 5
- * In this implementation M = 4.
- *
- * @param  {number} seed - seed. Example UTC milliseconds
- * @returns {string} - UUID
+ * Opens the survey URL in a new browser tab, appending parameters
+ * for the participant's survey ID (surveyID) and timezone offset
+ * (timezone).
  */
-function generateUUID(seed) {
-    return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
-        let r = Math.random() * 16;
-        r = (seed + r) % 16 | 0;
-        return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16);
-    });
-}
-
-async function openSurveyTab(useSameTab = false) {
-    const surveyId = await storageSpace.get("surveyId");
-    const surveyUrlFull = surveyUrlBase +
-        "?surveyId=" + surveyId +
-        "&timezone=" + new Date().getTimezoneOffset();
-    if (useSameTab) {
-        browser.tabs.update({url: surveyUrlFull});
-        return;
-    }
+async function openSurveyInNewTab() {
+    const surveyId = await getSurveyId();
+    const surveyUrlObj = new URL(surveyUrl);
+    surveyUrlObj.searchParams.append("surveyId", surveyId);
+    surveyUrlObj.searchParams.append("timezone", new Date().getTimezoneOffset());
     browser.tabs.create({
         active: true,
-        url: surveyUrlFull
+        url: surveyUrlObj.href
     });
 }
 
-async function requestSurvey(alarm) {
-    if (alarm.name == "surveyAlarm") {
-        const surveyCompleted = await storageSpace.get("surveyCompleted");
-        const noRequestSurvey = await storageSpace.get("noRequestSurvey");
-        if (surveyCompleted) return;
-        if (noRequestSurvey) return;
-        const currentTime = Date.now();
-        await storageSpace.set("lastSurveyRequest", currentTime);
-        browser.notifications.create({
-            type: "image",
-            message: surveyRequestMessage,
-            title: surveyRequestTitle,
-            iconUrl: SHIELD_URL
-        });
-        scheduleSurveyRequest(currentTime);
-    }
+/**
+ * Set a timeout to remind the user to complete the study.
+ */
+function scheduleReminderForUser() {
+    setTimeout(remindUser, Math.max((lastSurveyRequest + reminderInterval * millisecondsPerSecond) - Date.now(), 0));
```

Comment:
When `setTimeout` fires, it might be polite to the user to have `remindUser` called by `window.requestIdleCallback` but maybe I'm overthinking it.

Reply:
I _think_ it should be fine... `remindUser` immediately yields the event loop with an `await`.